eos-factory-reset-users-helper

```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+
+IFS=$'\n'
+for user_row in $(cat /etc/passwd); do
+    user=$(echo $user_row | cut -d : -f 1)
+    user_id=$(echo $user_row | cut -d : -f 3)
+
+    if [[ $user_id -ge 1000 && $user_id -le 1999 ]]; then
+        delete_tries=1
+        while true; do
+            echo "Deleting user $user ($user_id) (try: $delete_tries)"
+            userdel -r $user
+
+            if [ $? -eq 0 ]; then
+                break
+            else
+                if [ $delete_tries -gt 20 ]; then
+                    echo "Giving up on deleting $user ($user_id)"
+                    break
+                fi
+                sleep 2
+                delete_tries=$((delete_tries + 1))
```

Comment:
I don't think Rob's question at https://github.com/endlessm/eos-meta/pull/536#discussion_r233483330 is answered. I agree that it should not be necessary to have a retry loop.

Reply:
Try running `systemd-cgls` before the loop and see what the lingering processes are?

Reply:
I wonder if it might be enough to add `Conflicts=user@.service` – I don't know if you can use that syntax to match all instantiations of a template, but maybe you can.

Reply:
@wjt's idea is sensible, although we should probably look at either `user.slice` or the `user-<uid>.slice` for each existing user in the system. In any case, with our current logind configuration (inherited from Debian) we may wait forever if the user started a process in `screen`, for example. This can be solved changing `KillUserProcesses` in logind.conf.

But after giving some thought to all of this, I think ideally we should perform the factory reset in a more controlled environment. This way we can be sure no user processes are running nor will spawned (if the user logs in over ssh, for example), like during the next boot or during the shutdown process. On the factory testing process this is not problem because one of the requirements is that no users exist in the system.

Reply:
I agree that care should be taken to produce a highly deterministic environment under which this script is ran, where "nothing" is running except from this script. From a distance, I would have expected `systemctl isolate` to achieve this well enough:

>           This is similar to changing the runlevel in a traditional init
>           system. The isolate command will immediately stop processes that
>           are not enabled in the new unit, possibly including the graphical
>           environment or terminal you are currently using.

i.e. the concern about preventing users logging in over ssh should be a simple/natural part of that, since you would not include sshd in the factory-reset target being isolated here.

But if that's not the case (i.e. sshd and/or user session stuff is still running) then I also suggest going deeper into this so that we can understand what is happening and if `systemctl isolate` is or is not a suitable approach.